cron/workspacesLoad.js

```diff
@@ -0,0 +1,146 @@
+    },
+    'workspace_hosts_ready_count': {
+        'name': 'HostsReady',
+        'unit': 'Count',
+    },
+    'workspace_hosts_draining_count': {
+        'name': 'HostsDraining',
+        'unit': 'Count',
+    },
+    'workspace_hosts_unhealthy_count': {
+        'name': 'HostsUnhealthy',
+        'unit': 'Count',
+    },
+    'workspace_hosts_terminating_count': {
+        'name': 'HostsTerminating',
+        'unit': 'Count',
+    },
+    'workspace_hosts_stopped_count': {
+        'name': 'HostsStopped',
+        'unit': 'Count',
+    },
+    'workspaces_uninitialized_count': {
+        'name': 'WorkspacesUninitialized',
+        'unit': 'Count',
+    },
+    'workspaces_stopped_count': {
+        'name': 'WorkspacesStopped',
+        'unit': 'Count',
+    },
+    'workspaces_launching_count': {
+        'name': 'WorkspacesLaunching',
+        'unit': 'Count',
+    },
+    'workspaces_running_count': {
+        'name': 'WorkspacesRunning',
+        'unit': 'Count',
+    },
+    'workspaces_longest_launching_sec': {
+        'name': 'MaxLaunchingWorkspaceAge',
+        'unit': 'Seconds',
+    },
+    'workspaces_longest_running_sec': {
+        'name': 'MaxRunningWorkspaceAge',
+        'unit': 'Seconds',
+    },
+};
+async function sendStatsToCloudwatch(stats) {
+    const cloudwatch = new AWS.CloudWatch(config.awsServiceGlobalOptions);
+    const dimensions = [{Name: 'By Server', Value: config.workspaceCloudWatchName}];
+    const cloudwatch_metricdata_limit = 20; /* AWS limits to 20 items within each list */
+    const entries = Object.entries(stats).filter(([key, _value]) => {
+        return (key !== 'timestamp_formatted');
+    });
+
+    for (let i = 0; i < stats.length; i+= cloudwatch_metricdata_limit) {
+        const data = entries.slice(i, i + cloudwatch_metricdata_limit).map(([key, value]) => {
+            if (!(key in cloudwatch_definitions)) {
+                throw new Error(`Unknown datapoint ${key}!`);
+            }
+            const def = cloudwatch_definitions[key];
+            return {
+                MetricName: def.name,
+                Dimensions: dimensions,
+                Timestamp: stats.timestamp_formatted,
+                Unit: def.unit,
+                Value: value,
+            };
+        });
+        await cloudwatch.putMetricData({ Metricdata: data, Namespace: 'Workspaces' }).promise();
+    }
+}
+
+async function handleWorkspaceAutoscaling(stats) {
+    if (!(await config.getDBConfigValueAsync('workspaceAutoscalingEnabled', true))) return;
+
+    let desired_hosts = await config.getDBConfigValueAsync('workspaceDesiredHostCount');
+    if (!desired_hosts) {
+        desired_hosts = stats.workspace_hosts_desired;
+    }
+    let launch_template_id = await config.getDBConfigValueAsync('workspaceLaunchTemplateId');
+    if (!launch_template_id) {
+        launch_template_id = config.workspaceLoadLaunchTemplateId;
+    }
+
+    const running_hosts = stats.workspace_hosts_ready_count;
```

Comment:
I think we want the sum of `ready_count` and `launching_count`? Otherwise we will keep thinking we need to launch more, even though some are launching?

Reply:
Actually, I think this is a bit more subtle. I think that we want:
* if `desired_hosts > ready_hosts + launching_hosts` then start some more (or un-drain some)
* if `desired_hosts < ready_hosts` then set some running hosts to drain

I think we can't make this symmetric because:
* we don't want to use `desired_hosts > ready_hosts` for the scale-up check, because we will keep launching new hosts all the time while waiting for the launching ones to become ready.
* we don't want to use `desired_hosts < ready_hosts + launching_hosts`, because the launching ones don't really count yet for actual useable capacity